# Linear Algebra Using Python

This notebook demonstrates fundamental linear algebra concepts and operations using Python and NumPy.

## 1. Objectives

The objectives of this notebook are:

- Understand basic vector operations.
- Perform scalar multiplication.
- Calculate dot products and vector norms.
- Perform matrix addition and multiplication.
- Calculate matrix transpose, inverse, and rank.
- Calculate eigenvalues and eigenvectors.
- Interpret the results of linear algebra operations.

In [10]:
import numpy as np
import matplotlib.pyplot as plt

## 3. Vector Operations

Vectors are mathematical objects that contain an ordered collection of values. 
Python and NumPy can be used to perform different operations on vectors.

### 3.1 Vector Addition

Vector addition is performed by adding the corresponding elements of two vectors.

In [11]:
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])

vector_sum = vector_a + vector_b

print("Vector A:", vector_a)
print("Vector B:", vector_b)
print("Vector Sum:", vector_sum)

Vector A: [1 2 3]
Vector B: [4 5 6]
Vector Sum: [5 7 9]


**Interpretation:** The corresponding elements of the two vectors are added together. Therefore, the sum of `[1, 2, 3]` and `[4, 5, 6]` is `[5, 7, 9]`.

### 3.2 Scalar Multiplication

Scalar multiplication means multiplying every element of a vector by the same number.

In [12]:
vector = np.array([1, 2, 3])
scalar = 3

result = scalar * vector

print("Vector:", vector)
print("Scalar:", scalar)
print("Result:", result)

Vector: [1 2 3]
Scalar: 3
Result: [3 6 9]


**Interpretation:** Each element of the vector was multiplied by the scalar value 3. Therefore, `[1, 2, 3]` becomes `[3, 6, 9]`.

### 3.3 Dot Product and Norms

The dot product combines two vectors into a single numerical value. The norm measures the magnitude or length of a vector.

In [13]:
vector_a = np.array([1, 2, 3])
vector_b = np.array([4, 5, 6])

dot_product = np.dot(vector_a, vector_b)
norm_a = np.linalg.norm(vector_a)

print("Vector A:", vector_a)
print("Vector B:", vector_b)
print("Dot Product:", dot_product)
print("Norm of Vector A:", norm_a)

Vector A: [1 2 3]
Vector B: [4 5 6]
Dot Product: 32
Norm of Vector A: 3.7416573867739413


**Interpretation:** The dot product of the two vectors is 32. The norm of Vector A is approximately 3.74, which represents its magnitude.

## 4. Matrix Operations

Matrices are rectangular arrangements of numbers organized into rows and columns. NumPy provides several functions for performing matrix operations.

### 4.1 Matrix Addition and Multiplication

Matrix addition adds corresponding elements of two matrices. Matrix multiplication combines rows and columns according to matrix multiplication rules.

In [14]:
matrix_a = np.array([
    [1, 2],
    [3, 4]
])

matrix_b = np.array([
    [5, 6],
    [7, 8]
])

matrix_sum = matrix_a + matrix_b
matrix_product = matrix_a @ matrix_b

print("Matrix A:")
print(matrix_a)

print("\nMatrix B:")
print(matrix_b)

print("\nMatrix Addition:")
print(matrix_sum)

print("\nMatrix Multiplication:")
print(matrix_product)

Matrix A:
[[1 2]
 [3 4]]

Matrix B:
[[5 6]
 [7 8]]

Matrix Addition:
[[ 6  8]
 [10 12]]

Matrix Multiplication:
[[19 22]
 [43 50]]


### 4.2 Transpose and Inverse

The transpose changes rows into columns. The inverse of a square matrix is a matrix that can be used to obtain the identity matrix when multiplied by the original matrix.

In [15]:
matrix = np.array([
    [1, 2],
    [3, 4]
])

transpose = matrix.T
inverse = np.linalg.inv(matrix)

print("Original Matrix:")
print(matrix)

print("\nTranspose:")
print(transpose)

print("\nInverse:")
print(inverse)

Original Matrix:
[[1 2]
 [3 4]]

Transpose:
[[1 3]
 [2 4]]

Inverse:
[[-2.   1. ]
 [ 1.5 -0.5]]


### 4.3 Matrix Rank

The rank of a matrix represents the number of linearly independent rows or columns in the matrix.

In [16]:
matrix = np.array([
    [1, 2],
    [3, 4]
])

rank = np.linalg.matrix_rank(matrix)

print("Matrix:")
print(matrix)

print("\nRank:", rank)

Matrix:
[[1 2]
 [3 4]]

Rank: 2


## 5. Eigenvalues and Eigenvectors

### 5.1 Concept

Eigenvalues and eigenvectors are important concepts in linear algebra. For a square matrix, an eigenvector is a non-zero vector whose direction remains unchanged when the matrix is applied to it. The corresponding eigenvalue represents the amount of scaling.

### 5.2 Python Implementation

In [17]:
matrix = np.array([
    [2, 0],
    [0, 3]
])

eigenvalues, eigenvectors = np.linalg.eig(matrix)

print("Matrix:")
print(matrix)

print("\nEigenvalues:")
print(eigenvalues)

print("\nEigenvectors:")
print(eigenvectors)

Matrix:
[[2 0]
 [0 3]]

Eigenvalues:
[2.+0.j 3.+0.j]

Eigenvectors:
[[1.+0.j 0.+0.j]
 [0.+0.j 1.+0.j]]


### 5.3 Verification and Interpretation

The NumPy `linalg.eig()` function calculates the eigenvalues and corresponding eigenvectors of the matrix. The resulting eigenvalues describe the scaling factors associated with the eigenvectors.

## 6. Conclusion

This notebook demonstrated fundamental linear algebra operations using Python and NumPy. Vector addition, scalar multiplication, dot products, vector norms, matrix operations, matrix rank, eigenvalues, and eigenvectors were implemented.

The examples demonstrate how Python can be used to perform mathematical calculations efficiently while also providing clear numerical results. The combination of explanations, Python code, outputs, and interpretations provides a structured introduction to applying linear algebra concepts computationally.